command/init.go

```diff
@@ -767,6 +773,12 @@ func (c *InitCommand) getProviders(config *configs.Config, state *states.State,
 	// it's the smallest change relative to what came before it, which was
 	// a hidden JSON file specifically for tracking providers.)
 	if !newLocks.Equal(previousLocks) {
+		// if readonly mode, suppress changes
+		if flagLockfile == "readonly" {
+			log.Println("[DEBUG] init: detected changing dependencies, but suppressed by readonly mode")
```

Comment:
I think we may want to output a warning to the UI here rather than a debug log. Maybe returning a warning diagnostic? For example:

```go
diags = diags.Append(tfdiags.Sourceless(
  tfdiags.Warning,
  "Provider lock file not updated",
  "Changes to the provider selections were detected, but not saved in the .terraform.lock.hcl file. To record these selections, run "terraform init" without the "-lockfile=readonly" flag.,
))
```

Reply:
Fixed.